l10n_it_fiscalcode/wizard/compute_fc.py

```diff
@@ -1,54 +1,110 @@
 # Copyright 2014 Associazione Odoo Italia (<http://www.odoo-italia.org>)
 # Copyright 2016 Andrea Gallina (Apulia Software)
+# Copyright © 2018 Matteo Bilotta (Link IT s.r.l.)
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
+import datetime
+import logging
+
 from odoo import models, fields, api, _
 from odoo.exceptions import UserError
-import logging
-import datetime
 
 _logger = logging.getLogger(__name__)
 
 try:
     from codicefiscale import build
+
 except ImportError:
     _logger.warning(
-        'codicefiscale library not found. '
-        'If you plan to use it, please install the codicefiscale library '
-        'from https://pypi.python.org/pypi/codicefiscale')
+        "codicefiscale library not found. "
+        "If you plan to use it, please install the codicefiscale library"
+        " from https://pypi.python.org/pypi/codicefiscale")
 
 
 class WizardComputeFc(models.TransientModel):
-
-    _name = "wizard.compute.fc"
+    _name = 'wizard.compute.fc'
     _description = "Compute Fiscal Code"
     _rec_name = 'fiscalcode_surname'
 
-    fiscalcode_surname = fields.Char('Surname', size=64)
-    fiscalcode_firstname = fields.Char('First name', size=64)
-    birth_date = fields.Date('Date of birth')
-    birth_city = fields.Many2one(
-        'res.city.it.code.distinct', string='City of birth')
-    birth_province = fields.Many2one(
-        'res.city.it.code.province', string='Province')
-    sex = fields.Selection([
-        ('M', 'Male'),
-        ('F', 'Female'),
-        ], "Sex")
+    fiscalcode_surname = fields.Char("Surname", required=True, size=64)
+    fiscalcode_firstname = fields.Char("First name", required=True, size=64)
+    birth_date = fields.Date("Date of birth", required=True)
+    birth_city = fields.Many2one('res.city.it.code.distinct',
+                                 required=True,
+                                 string="City of birth")
+    birth_province = fields.Many2one('res.country.state',
```

Comment:
@Byloth usi ```res.country.state``` assumendo che i dati siano stati creati con ```base_location_geonames_import```, giusto?

Reply:
Ciao!
No, ad essere sincero no.

Il modello `res.country.state` viene popolato tramite un apposito [file CSV](https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/res/res.country.state.csv#L307) presente all'interno del modulo `base` di Odoo.

Per essere ulteriormente precisi, le righe che interessano a noi di quel file, sono solo 111 *(banalmente, le "province" straniere non ci interessano, per la generazione del codice fiscale Italiano)*.

---

**Off topic**

Il modulo che tu proponi, in effetti, sembra interessante per andare a sostituire la vista `res_city_it_code_distinct` contenente, di fatto, i nomi di tutte le città Italiane catalogate.

Ho trovato delle differenze *importanti* tra il numero delle righe presenti in questa view e il numero delle righe presenti all'interno dello ZIP usato durante la [procedura di configurazione](https://github.com/OCA/partner-contact/tree/11.0/base_location_geonames_import#configuration) indicata dal modulo che hai suggerito: 11.821 contro 18.377.  
Non ho analizzato nello specifico da cosa possa essere dovuta questa differenza di 6.556 righe *(vere città effettive?)* ma potrebbe valere la pena spenderci un po' di tempo...

Reply:
Non so però se l'utilizzo di  ```base_location_geonames_import``` possa sostituire ```res_city_it_code_distinct``` perchè, se non sbaglio, quest'ultima contiene anche i codici catastali, utilizzati per il calcolo del codice fiscale, mentre fra i dati caricati da ```base_location_geonames_import``` non mi pare ci siano

Reply:
Mmmh... 🤔  
Non ho analizzato nel dettaglio la questione; boh... Potrebbe essere!